### PR TITLE
Add Ens Client

### PR DESCRIPTION
**Pull request by falvaradorodriguez**
closes https://github.com/safe-global/safe-eth-py/issues/754

### CODE REVIEW

File: gnosis/eth/clients/ens_client.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from typing import Any, Dict, List, Optional, Union
+
+import requests
+from cache_memoize import cache_memoize
+from hexbytes import HexBytes
+
+from gnosis.eth import EthereumNetwork
+
+
+class EnsClient:
```
Comment (Uxio0) [inline gnosis/eth/clients/ens_client.py:11]:
```suggestion
class EnsClient:
"""
Resolves Ethereum Name Service domains using ``thegraph`` API
"""
```

File: gnosis/eth/clients/ens_client.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            url = (
+                "https://api.studio.thegraph.com/proxy/49574/enssepolia/version/latest/"
+            )
+        else:  # Fallback to mainnet
+            url = "https://api.thegraph.com/subgraphs/name/ensdomains/ens/"
+        self.url = url
+        self.request_timeout = 5  # Seconds
+        self.request_session = requests.Session()
+
+    def is_available(self):
```
Comment (Uxio0) [inline gnosis/eth/clients/ens_client.py:24]:
```suggestion
    def is_available(self) -> bool:
```

File: gnosis/eth/clients/ens_client.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        """
+        :return: True if service is available, False if it's down
+        """
+        try:
+            return self.request_session.get(self.url, timeout=self.request_timeout).ok
+        except IOError:
+            return False
+
+    @staticmethod
+    def domain_hash_to_hex_str(domain_hash: Union[str, bytes, int]) -> str:
```
Comment (Uxio0) [inline gnosis/eth/clients/ens_client.py:34]:
```suggestion
    def domain_hash_to_hex_str(domain_hash: Union[str, bytes, int]) -> HexStr:
```

File: gnosis/eth/clients/ens_client.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        if response.ok:
+            data = response.json()
+            if data:
+                domains = data.get("data", {}).get("domains")
+                if domains:
+                    return domains[0].get("labelName")
+        return None
+
+    def query_by_domain_hash(
+        self, domain_hash: Union[str, bytes, int]
```
Comment (Uxio0) [inline gnosis/eth/clients/ens_client.py:83]:
```suggestion
        self, domain_hash: Union[HexStr, bytes, int]
```

File: gnosis/eth/clients/ens_client.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        :param domain_hash:
+        :return: Domain hash as an hex string of 66 chars (counting with 0x), padding with zeros if needed
+        """
+        if not domain_hash:
+            domain_hash = b""
+        return "0x" + HexBytes(domain_hash).hex()[2:].rjust(64, "0")
+
+    @lru_cache
+    @cache_memoize(60 * 60 * 24, prefix="ens-_query_by_domain_hash")  # 1 day
+    def _query_by_domain_hash(self, domain_hash_str: str) -> Optional[str]:
```
Comment (Uxio0) [inline gnosis/eth/clients/ens_client.py:45]:
```suggestion
    def _query_by_domain_hash(self, domain_hash_str: HexStr) -> Optional[str]:
```

File: requirements.txt [hunk 6]
```diff
@@ -1,4 +1,5 @@
 django==4.2.10
+django-cache-memoize==0.2.0
```
Comment (Uxio0) [inline requirements.txt:2]: Don't use cache-memoize, standard library should be enough

Comment (falvaradorodriguez) [reply requirements.txt:2]: Ok, is `gnosis.util.cache` correct here?

Comment (Uxio0) [reply requirements.txt:2]: Yes, it's a wrapper around `functools.cache` as it was added in Python 3.9 : https://docs.python.org/3/library/functools.html#functools.cache

**Review by Uxio0 — APPROVED** [review]

**Pull request merged**